pkg/collector/process.go

```diff
@@ -0,0 +1,266 @@
+		}
+		indicator := elementID[0] >> 7
+		if indicator != 1 {
+			element.elementID = binary.BigEndian.Uint16(elementID)
+			element.enterpriseID = uint32(0)
+			element.elementLength = elementLength
+		} else {
+			err = decode(templateBuffer, &element.enterpriseID)
+			if err != nil {
+				return nil, fmt.Errorf("Error in decoding message: %v", err)
+			}
+			elementID[0] = elementID[0] ^ 0x80
+			element.elementID = binary.BigEndian.Uint16(elementID)
+			element.elementLength = elementLength
+		}
+		templateMsg.AddInfoElement(element.enterpriseID, element.elementID)
+		elements = append(elements, &element)
+	}
+	cp.addTemplate(obsDomainID, templateID, elements)
+	return templateMsg, nil
+}
+
+func (cp *collectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID uint32, templateID uint16) (interface{}, error) {
+	// make sure template exists
+	template, err := cp.getTemplate(obsDomainID, templateID)
+	if err != nil {
+		return nil, fmt.Errorf("Template %d with obsDomainID %d does not exist", templateID, obsDomainID)
+	}
+	dataMsg := entities.NewDataSet()
+	for _, field := range template {
+		var length int
+		if field.elementLength == entities.VariableLength { // string
+			length = getFieldLength(dataBuffer)
+		} else {
+			length = int(field.elementLength)
+		}
+		val := dataBuffer.Next(length)
+		dataMsg.AddInfoElement(field.enterpriseID, field.elementID, val)
+	}
+	return dataMsg, nil
+}
+
+func (cp *collectingProcess) addTemplate(obsDomainID uint32, templateID uint16, elements []*templateField) {
+	cp.templatesLock.Lock()
+	if _, exists := cp.templatesMap[obsDomainID]; !exists {
+		cp.templatesMap[obsDomainID] = make(map[uint16][]*templateField)
+	}
+	cp.templatesMap[obsDomainID][templateID] = elements
+	cp.templatesLock.Unlock()
+	// template lifetime management
+	if cp.address.Network() == "tcp" {
+		return
+	}
+
+	// Handle udp template expiration
+	if cp.templateTTL == 0 {
+		cp.templateTTL = config.TemplateTTL // Default value
+	}
+	go func() {
+		ticker := time.NewTicker(time.Duration(cp.templateTTL) * time.Second)
+		defer ticker.Stop()
+		select {
+		case <-ticker.C:
+			klog.Infof("Template with id %d, and obsDomainID %d is expired.", templateID, obsDomainID)
+			cp.deleteTemplate(obsDomainID, templateID)
+			break
+		}
+	}()
+}
+
+func (cp *collectingProcess) getTemplate(obsDomainID uint32, templateID uint16) ([]*templateField, error) {
+	cp.templatesLock.RLock()
+	defer cp.templatesLock.RUnlock()
+	if elements, exists := cp.templatesMap[obsDomainID][templateID]; exists {
+		return elements, nil
+	} else {
+		return nil, fmt.Errorf("Template %d with obsDomainID %d does not exist.", templateID, obsDomainID)
+	}
+}
+
+func (cp *collectingProcess) deleteTemplate(obsDomainID uint32, templateID uint16) {
+	cp.templatesLock.Lock()
+	defer cp.templatesLock.Unlock()
+	delete(cp.templatesMap[obsDomainID], templateID)
+}
+
+// a common binary decoder to specified interfaces
```

Comment:
Convention for golang says that comment should start with function name.
https://blog.golang.org/godoc